docs/_posts/2019-11-14-wash-gem.md

```diff
@@ -0,0 +1,132 @@
+
+{% highlight ruby %}
+class Puppetwash < Wash::Entry
+  label 'puppet'
+  is_singleton
+  parent_of 'PEInstance'
+
+  def init(_wash_config)
+  end
+
+  def list
+    config.keys.map do |name|
+       PEInstance.new(name)
+    end
+  end
+end
+
+class PEInstance < Wash::Entry
+  label 'pe_instance'
+  parent_of 'NodesDir'
+
+  def initialize(name)
+    @name = name
+  end
+
+  def list
+    [NodesDir.new('nodes', name)]
+  end
+end
+
+# Report relies on end_time and hash. The others are included as useful metadata.
+METADATA_FIELDS = {
+  'end_time': 'string',
+  'environment': 'string',
+  'status': 'string',
+  'noop': 'boolean',
+  'puppet_version': 'string',
+  'producer': 'string',
+  'hash': 'string'
+}
+
+class Report < Wash::Entry
+  label 'report'
+  attributes :meta, :mtime
+  meta_attribute_schema(
+      type: 'object',
+      properties: METADATA_FIELDS.map { |k, v| [k, { type: v }] }.to_h
+  )
+  state :node_name, :pe_name, :hash
+
+  def initialize(report, node_name, pe_name)
+    @name = report['end_time']
+    @node_name = node_name
+    @pe_name = pe_name
+    @hash = report['hash']
+    @meta = report
+    @mtime = Time.parse(report['end_time'])
+  end
+
+  def read
+    response = client(@pe_name).request(
+      'reports',
+      [:and, [:'=', :certname, @node_name], [:'=', :hash, @hash]]
+    )
+    make_readable(response.data)
+  end
+end
+{% endhighlight %}
+
+Notice that
+
+* Each class extends the `Wash::Entry` class.
+* Entry schemas are declared via class-level helpers like `label`, `is_singleton`, and `parent_of`.
+* Each Wash method corresponds to an instance method in the class. For example, a report's `read` action is implemented by `Report#read`. Similarly, a PE instance's `list` action is implemented by `PEInstance#list`.
+* Entry attributes are declared via the `attributes` helper and can be set/accessed as instance-level fields. For example, a report's `mtime` and `meta` attributes are set in the constructor (`initialize` method).
+* The `state` helper lets you specify instance-level fields that should be serialized as part of the `state` key in the entry's JSON. The `wash gem` will set these fields after it reconstructs your entry. Thus, you can safely access those fields' values when implementing the entry's methods. For example, the `@pe_name` field in `Report` is initialized in the constructor and then referenced in `Report#read`.
+
+The `Wash::Entry` class makes all of this possible. It also comes with even more helpers that let you do other things like specify cache TTLs (see `Wash::Entry#cache_ttls`) or any prefetched methods (see `Wash::Entry#prefetch`).
+
+Recall that plugin script invocations adopt the following calling convention
+
+```
+<plugin_script> <method> <entry> <args...>
+```
+
+All plugin script invocations are handled by the `Wash.run` function. Here's how `Wash.run` is used in `puppetwash`
```

Comment:
I'd call out more clearly that `puppetwash` here refers to the script, and that the call to `Wash.run` is added at the end of it to actually run things. I don't feel like that's quite clear from what you've said here.